### ISSUE TITLE
Support SSL/TLS for MySQL/MariaDB secrets engine

### ISSUE THREAD

**Issue by bradjones1** [hashicorp/vault#6444]
**Is your feature request related to a problem? Please describe.**

The MySQL storage engine supports this as of https://github.com/hashicorp/vault/pull/439 - there is however no way currently to [configure a TLS configuration](https://github.com/go-sql-driver/mysql#tls) for use in a connection DSN for the secrets engine.

**Describe the solution you'd like**

A PR very similar to that used for the storage engine could be applied to the MySQL secrets engine.

**Describe alternatives you've considered**

A user may be able to proxy through to the server via non-SSL means, however this is not possible in all environments.

**Explain any additional use-cases**

Our use case stems from the necessity to employ first-generation Google Cloud SQL instances which do not support GCP's encryption proxy. However, this use case is not unique to Google Cloud at all; SSL for MySQL is supported widely across flavors and hosting providers.

**Comment by rorynscott** [comment]
@bradjones1 were you ever able to solve this issue? We are running into what I think is the same issue where our Cloud SQL MySQL instances require SSL and we're having a heck of a time trying to get the vault configuration to use the certs to connect. We were able to pass Postgres the certificate path in the connection string, but everything we've tried to far with MySQL has failed.

**Comment by bradjones1** [comment]
@rorynscott Short answer: no. Slightly longer answer: We are on GCP and first-gen instances are sunset. TLS/SSL is still an option to connect to second-gen but there is also a very good proxy container/service that Google provides to tunnel, so the short answer there is to run the proxy in a pod or other trusted network plane with Vault and then let the proxy handle the encryption over the wire. So this has basically become a non-issue for us.

I suppose in theory you could also use something like https://proxysql.com/ in other environments as a roll-your-own similar to what I described above. We use that for non-administrative access from outside our network and aside from the crazy CPU utilization in clustered mode, it works really well.

tl;dr: this is still an outstanding request but not one that is blocking us (not that it matters to the contributors!) but there are some ideas for you.

**Comment by rorynscott** [comment]
@bradjones1 thanks for the suggestion. I toyed around with the proxy as a sidecar to vault yesterday and it does look like that'll work for us. Not sure if it's ideal or not, but working software is great. Thanks again!

**Comment by narayan-iyengar** [comment]
Is the issue talking about implementing mTLS between Vault and MySQL?

**Comment by bradjones1** [comment]
@narayan-iyengar Yes - Basically SSL configuration is not exposed by vault for these connections.

**Comment by pcman312** [comment]
@bradjones1 Are you using an mTLS connection to MySQL? Or are you just using a custom CA that the MySQL server's certificate was issued from?

**Comment by bradjones1** [comment]
@pcman312 It would have been mTLS, however we are no longer needing to talk to these services from Vault any longer. We're able to utilize a sidecar method (e.g., what I recommended at https://github.com/hashicorp/vault/issues/6444#issuecomment-621302464 ) so my interest here is purely academic at this point.

**Comment by rorynscott** [comment]
This seems like something Vault would want to offer, but we, too, switched to using the GCP CloudSQL Proxy as a sidecar to vault.

**Comment by Valarissa** [comment]
#9181 has been created to address this issue.

**Comment by omelnyk1** [comment]
@bradjones1 I faced the same issue. It was fixed by adding NAT GW IP addresses to authorize networks 
 in SQL instance settings https://cloud.google.com/sql/docs/mysql/authorize-networks

**Comment by CPCJ79** [comment]
I need this to work, we have sensitive database end points that we do not proxy by policy, we terminate directly.  This means I would need the ability to specify the SSL_MODE option to VERIFY_CA as the standard mysql tools do not issue tls certificates with hostnames in them, there is merely an issuer trust verification.  Right now, I can not specify VERIFY_CA or VERIFY_IDENTITY depending on a publicly valid TLS certificate with a dns name that can be verified : 

https://dev.mysql.com/doc/refman/8.0/en/connection-options.html#option_general_ssl-mode

**Comment by bradjones1** [comment]
I'll just chime in here to note, I'm sure Hashicorp would be happy to talk to those who "need" this about sponsored development or a paid support plan for Vault. This being open source software, of course, there's no guarantee of any particular feature set if you aren't a paying customer. Those of us who use Vault open source are very appreciative of the work Hashicorp has put into building it.
